doc/src/parsing_idioms.md

```diff
@@ -0,0 +1,190 @@
+# grmtools parsing idioms
+
+grmtools is a flexible tool and can be used in many ways. However, for those
+using the `Grmtools` format, the simple idioms can often make life easier.
+
+
+# Return `Lexeme`s rather than `&str`s or `String`s when performance is critical
+
+When executing grammar actions one is often building up an Abstract Syntax Tree
+(AST) or equivalent. For example consider a simple language with assignments:
+
+```ignore
+Assign: "ID" "=" Expr;
+```
+
+Perhaps the "obvious" way to build this into an AST is to extract the string
+representing the identifier as follows:
+
+```
+Assign -> ASTAssign: "ID" "+" Expr
+    {
+        let id = $lexer.lexeme_str($1.as_ref().unwrap()).to_string();
+        ASTAssign::new(id, $3)
+    }
+
+%%
+
+struct ASTAssign {
+    id: String
+}
+
+impl ASTAssign {
+    fn new(name: String) -> Self {
+        ASTAssign { name }
+    }
+}
+```
+
+This approach is easy to work with, but isn't as performant as may be desired:
+the `to_string` call allocates memory and copies part of the user's input into
+that.
+
+An alternative approach is not to convert the lexeme into a `String` during
+parsing, but simply to return the
+[`Lexeme`](https://docs.rs/lrpar/~0/lrpar/lex/struct.Lexeme.html) itself. The
+relevant `&str` slice can be extracted from the user's input later and memory
+allocation avoided entirely. An example of this is as follows:
+
+```
+Assign -> ASTAssign: "ID" "+" Expr
+    {
+        let id = $lexer.lexeme_str($1);
+        ASTAssign::new(id, $3)
+    }
+
+%%
+
+type StorageT = u32;
+
+struct ASTAssign {
+    id: Lexeme<StorageT>
+}
+
+impl ASTAssign {
+    fn new(name: Lexeme<StorageT>) -> Self {
+        ASTAssign { name }
+    }
+}
+```
+
+Dealing with the [`Lexeme`](https://docs.rs/lrpar/~0/lrpar/lex/struct.Lexeme.html)
+type can be somewhat fiddly because it has a type paramater `<StorageT>`
+with somewhat complex bounds. In essence, `StorageT` must be an unsigned integer
+that is big enough to index (individually) a grammar's rules, productions, and
+lexemes. Users can manually specify `StorageT` by using
+[`CTParserBuilder::new_with_storaget`](https://docs.rs/lrpar/0.1.1/lrpar/ctbuilder/struct.CTParserBuilder.html#method.new_with_storaget);
+if not specified explicitly it defaults to `u32`.
+
+A simple way of avoiding the complexity of the parameter's bounds is to encode
+it directly into the `Lexeme` type. A pleasant, and easily adapted way of doing
+this, is to semi-hard-code the value of `StorageT` into the AST, which we
+do by adding `type StorageT = u32;` to the programs part of the grammar
+(if [`CTParserBuilder::new_with_storaget`](https://docs.rs/lrpar/0.1.1/lrpar/ctbuilder/struct.CTParserBuilder.html#method.new_with_storaget)
```

Comment:
Sentence in brackets doesn't parse. Maybe a missing `or` before `if`?

Reply:
Oops, there's a spurious ')' in there.